File: Easy_Problems/JewelsAndStones.py
"""
LeetCode Problem: Jewels and Stones
Link: https://leetcode.com/problems/jewels-and-stones/

You're given strings `jewels` and `stones`. Each character in `stones` represents a stone you have.
Each character in `jewels` represents a type of stone that is a jewel.
Return the number of stones that are also jewels.

Letters are case-sensitive, so "a" and "A" are considered different.

Examples:
  Input: jewels = "aA", stones = "aAAbbbb"
  Output: 3

  Input: jewels = "z", stones = "ZZ"
  Output: 0

Approach:
  - Add each character from `jewels` to a set for O(1) lookup
  - Iterate through `stones` and count how many are in the set

Time Complexity: O(n + m) — n = len(jewels), m = len(stones)
Space Complexity: O(n) — for storing the jewel set
Tags: Hash Set, String, Counting
"""

class Solution:
    def numJewelsInStones(self, jewels: str, stones: str) -> int:
        jewelSet = set()
        count = 0

        # Add each jewel type to the set
        for ch in jewels:
            jewelSet.add(ch)

        # Count stones that are jewels
        for ch in stones:
            if ch in jewelSet:
                count += 1

        return count

File: Easy_Problems/JewelsAndStones.cpp
/*
LeetCode Problem: Jewels and Stones
Link: https://leetcode.com/problems/jewels-and-stones/

You're given two strings: `jewels` (representing types of jewels) and `stones`
(representing the stones you have). Each character in `stones` is a type of
stone, and you want to count how many are jewels.

Letters are case-sensitive. All characters in `jewels` are guaranteed unique.

Examples:
  Input:  jewels = "aA", stones = "aAAbbbb"
  Output: 3  → (a, A, A)

  Input:  jewels = "z", stones = "ZZ"
  Output: 0

Approach:
  - Store each jewel type in a set for O(1) lookups.
  - Loop through each character in `stones` and increment a counter
    whenever that stone is also in the set.

Time Complexity: O(m + n) — m = length of jewels, n = length of stones
Space Complexity: O(m) — for storing jewel types
Tags: Hash Table, String, Set, Frequency Count
*/

class Solution {
public:
    int numJewelsInStones(string jewels, string stones) {
        set<char> jewelSet;
        int count = 0;

        // Populate set with jewel types
        for (int i = 0; i < jewels.length(); i++) {
            jewelSet.insert(jewels[i]);
        }

        // Count how many stones are jewels
        for (int i = 0; i < stones.length(); i++) {
            if (jewelSet.count(stones[i])) {
                count++;
            }
        }

        return count;
    }
};